cachelib/shm/FileShmSegment.cpp

```diff
@@ -0,0 +1,341 @@
+  }
+}
+
+} // namespace detail
+
+FileShmSegment::FileShmSegment(ShmAttachT,
+                                 const std::string& name,
+                                 ShmSegmentOpts opts)
+    : ShmBase(std::move(opts), name),
+      fd_(getExisting(getPath(), opts_)) {
+  XDCHECK_NE(fd_, kInvalidFD);
+  markActive();
+  createReferenceMapping();
+}
+
+FileShmSegment::FileShmSegment(ShmNewT,
+                                 const std::string& name,
+                                 size_t size,
+                                 ShmSegmentOpts opts)
+    : ShmBase(std::move(opts), name),
+      fd_(createNewSegment(getPath())) {
+  markActive();
+  resize(size);
+  XDCHECK(isActive());
+  XDCHECK_NE(fd_, kInvalidFD);
+  // this ensures that the segment lives while the object lives.
+  createReferenceMapping();
+}
+
+FileShmSegment::~FileShmSegment() {
+  try {
+    // delete the reference mapping so the segment can be deleted if its
+    // marked to be.
+    deleteReferenceMapping();
+  } catch (const std::system_error& e) {
+  }
+
+  // need to close the fd without throwing any exceptions. so we call close
+  // directly.
+  if (fd_ != kInvalidFD) {
+    const int ret = close(fd_);
+    if (ret != 0) {
+      XDCHECK_NE(errno, EIO);
+      XDCHECK_NE(errno, EINTR);
+      XDCHECK_EQ(errno, EBADF);
+      XDCHECK(!errno);
+    }
+  }
+}
+
+int FileShmSegment::createNewSegment(const std::string& name) {
+  constexpr static int createFlags = O_RDWR | O_CREAT | O_EXCL;
+  return detail::openImpl(name.c_str(), createFlags);
+}
+
+int FileShmSegment::getExisting(const std::string& name,
+                                 const ShmSegmentOpts& opts) {
+  int flags = opts.readOnly ? O_RDONLY : O_RDWR;
+  return detail::openImpl(name.c_str(), flags);
+}
+
+void FileShmSegment::markForRemoval() {
+  if (isActive()) {
+    // we still have the fd open. so we can use it to perform ftruncate
+    // even after marking for removal through unlink. The fd does not get
+    // recycled until we actually destroy this object.
+    removeByPath(getPath());
+    markForRemove();
+  } else {
+    XDCHECK(false);
+  }
+}
+
+bool FileShmSegment::removeByPath(const std::string& path) {
+  try {
+    detail::unlinkImpl(path.c_str());
+    return true;
+  } catch (const std::system_error& e) {
+    // unlink is opaque unlike sys-V api where its through the shmid. Hence
+    // if someone has already unlinked it for us, we just let it pass.
+    if (e.code().value() != ENOENT) {
+      throw;
+    }
+    return false;
+  }
+}
+
+std::string FileShmSegment::getPath() const {
+  return std::get<FileShmSegmentOpts>(opts_.typeOpts).path;
+}
+
+size_t FileShmSegment::getSize() const {
+  if (isActive() || isMarkedForRemoval()) {
+    stat_t buf = {};
+    detail::fstatImpl(fd_, &buf);
+    return buf.st_size;
+  } else {
+    throw std::runtime_error(folly::sformat(
+        "Trying to get size of  segment with name {} in an invalid state",
```

Comment:
""Trying to get size of  segment" -> ""Trying to get size of segment" (delete extra space).